Link Preview block

notnownikki opened the pull request:
## Description

Adds a "Link Preview" block, in the style of the preview that Facebook gives when you put a URL in a status update.

If the Embed block fails to embed a URL, it falls back to the Link Preview block. The block can be used separately too, as some pages look better using the OpenGraph content than they do with what oEmbed returns.

If this is not wanted, then a fallback to a paragraph block would give us parity with the current editor for URLs that can't be embedded.

A demo! This shows the block being used as an oEmbed fallback (which gives a preview block and a paragraph block with a link in it, for maximum user options!) and using the preview block as a standalone thing.

![preview](https://user-images.githubusercontent.com/3314354/41677756-2dd3156a-74c1-11e8-8512-3512c186cd75.gif)


There's a new API endpoint. Why a new API?

I wanted to cover these use cases:

1) User wants to embed a URL not supported by oEmbed
2) User does not want the oEmbed content, just wants a preview of the link instead

We want to be able to give information about a URL, either using the OpenGraph metadata,
or using content from the page, to generate a preview for the link.

The user might want to make changes to this preview before publishing. For example,
if we're generating preview data from the content of the page, we'll want to display
a range of images so that the user can choose which one best represents the page
they're linking to. This will happen when the page doesn't have OpenGraph metadata.
The user might want to remove the image altogether, if there's nothing suitable there.

This means that we can't just have a HTML fallback, using the `embed_maybe_make_link` filter.
We want to be able to return data that the Link Preview block will use in its editing
UI, and what the user chooses there will determine the HTML saved.

So I've considered the following approaches:

HTML fallback - not suitable because we need data about the page we're previewing so
the user can make a decision about what HTML ends up in their post.

Inject data into `embed/proxy` response - this would change the signature of the
response in an undocumented way, and also either introduce an extra HTTP request
for every embed proxy request if we presented the data with every embed call,
or render the user unable to use the link preview as an alternative to the
embed block if we only generated the data for URLs that oEmbed did not support.

New API - Means an extra API call if oEmbed does not support the URL, but gives
us the flexibility in editing and allowing the user to choose a preview instead
of an embed. This gives us more future flexibility to create link preview blocks
for links that appear in paragraphs, in the same way Facebook do when you put
a URL in a status update.

## How has this been tested?

Test with:

1) A URL with OpenGraph information (e.g. https://tabs.ultimate-guitar.com/tab/misc_cartoons/steven_universe_-_whats_the_use_of_feeling_blue_ukulele_2155897 )
2) A URL with no OpenGraph information, and a range of images to choose from (e.g. https://wordpress.org/ )
3) A URL that would be embeddable with oEmbed (e.g. https://css-tricks.com/snippets/css/a-guide-to-flexbox/ )


## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

Soean commented on core-blocks/link-preview/index.js:
```diff
@@ -0,0 +1,242 @@
+						</div>
+						<div className="wp-block-link-preview__image__tools">
+							{ hasMultipleImages && <Button onClick={ this.previousImage }>&lt;</Button> }
+							<Button onClick={ this.removeImage }>remove</Button>
```
`remove` should be translatable

Soean commented on lib/class-wp-rest-opengraph-controller.php:
```diff
@@ -0,0 +1,202 @@
+				'permission_callback' => array( $this, 'get_preview_permissions_check' ),
+				'args'                => array(
+					'url' => array(
+						'description'       => __( 'The URL of the resource for which to fetch oEmbed data.' ),
```
Text domain is missing

Soean commented on lib/class-wp-rest-opengraph-controller.php:
```diff
@@ -0,0 +1,202 @@
+		if ( ! current_user_can( 'edit_posts' ) ) {
+			return new WP_Error(
+				'rest_forbidden',
+				__( 'Sorry, you are not allowed to make OpenGraph requests.' ),
```
Text domain is missing

Soean commented on lib/class-wp-rest-opengraph-controller.php:
```diff
@@ -0,0 +1,202 @@
+	/**
+	 * Constructs the controller.
+	 *
+	 * @access public
```
No need for access, it's already in the code.

Soean commented on lib/class-wp-rest-opengraph-controller.php:
```diff
@@ -0,0 +1,202 @@
+		return $data;
+	}
+
+	private function generate_preview( $args ) {
```
Docs are missing

Soean commented:
Nice block! 
Maybe we should filter out small images, which are smaller than 100px?

Soean commented on core-blocks/link-preview/index.js:
```diff
@@ -0,0 +1,242 @@
+							<img src={ selectedImage.src } alt="" />
+						</div>
+						<div className="wp-block-link-preview__image__tools">
+							{ hasMultipleImages && <Button onClick={ this.previousImage }>&lt;</Button> }
```
Use Dashicons instead? 
https://developer.wordpress.org/resource/dashicons/#arrow-left-alt2

jasmussen commented:
This is really impressive! Quite cool!

GIF:

![linkpreview](https://user-images.githubusercontent.com/1204802/41582747-b896fefc-73a2-11e8-8f23-3e8e008094bb.gif)

I don't know why the next/prev buttons didn't quite work, they worked fine the first time. In any case, this is very promising.

- I think we should tweak the UI for picking the thumbnail a bit. They could be IconButtons with a next and prev dashicon inside. 
- This UI should also only be visible when the block is selected. 


I also think it's pretty key that you can paste a link on a new line, and if it the link fails to embed, it outputs _two_ blocks, if that's even technically possible. First block should just be a paragraph block with the link you pasted, and the 2nd could then be the link preview block. Like this:

<img width="750" alt="screen shot 2018-06-19 at 09 28 22" src="https://user-images.githubusercontent.com/1204802/41582925-3dc2e870-73a3-11e8-9273-c536a97051a9.png">

The point is — this would allow the user to have it their own way. Either they want just the link, just the link block, or hey, both. By outputting two blocks, the user can delete the one they don't want.

notnownikki commented:
What URL did you test with there? I think the reason the arrows appeared not too work is because the scraping picks up tracking images, which are transparent, so you're cycling through them... But they're invisible

jasmussen commented:
https://wordpress.org

I also think I was a bit quick — one Accessibility icon was picked up after I waited a bit :D 

Maybe only pick images > n pixels? Or just not scrape for now, and rely only on opengraph data, that could work too.

notnownikki commented:
As suggested by a few people now, I'm going to make it only pick out images over a certain size. We'll generate thumbnails of those images in the media library so we're not hotlinking, and not using larger images than we need to.

Also going to get Dashicons in those buttons, and have the image selection / removal controls only show then the block is focused.

notnownikki commented:
BTW, if we decide that disabling the scraping and only supporting OpenGraph is the way to go, it's easy to do, it's just removing some code. The block doesn't care where the data comes from, it just uses what the API gives it.

notnownikki commented:
Making progress...

Using IconButtons for the image selection controls.
Image controls only visible when block is selected.
Preview images are served from the media library instead of the remote site, and resized to thumbnails before upload so we're not storing the full size images.
Filter out potential preview images by size.

jasmussen commented on core-blocks/embed/index.js:
```diff
@@ -176,7 +176,12 @@ function getEmbedBlockSettings( { title, description, icon, category = 'embed',
-								this.props.onReplace( createBlock( 'core-embed/link-preview', { url } ) );
+								this.props.onReplace(
+									[
+										createBlock( 'core/paragraph', { content: <a href={ url }>{ url }</a> } ),
```
😍

notnownikki commented:
And now, as @jasmussen suggested, when the link preview block is used as a fallback for the embed block, it gets a paragraph with a link to the URL in too, as shown in https://github.com/WordPress/gutenberg/pull/7350#issuecomment-398302499

jasmussen commented:
🤟

notnownikki commented:
I think this is ready for a review now, as a first iteration.

In a future iteration, I'd like to be able to replace the image with one from the media library, but I think this  version would be a good first release.

Soean commented on core-blocks/link-preview/index.js:
```diff
@@ -0,0 +1,252 @@
+						{ isSelected &&
+						<div className="wp-block-embed-link-preview__image__tools">
+							{ hasMultipleImages && <IconButton aria-label="Previous image" icon="arrow-left-alt2" onClick={ this.previousImage } /> }
+							<IconButton icon="no" aria-label="Remove image" onClick={ this.removeImage } />
```
We should also translate the aria-labels

notnownikki replied:
As I was typing them, I was thinking, "these need to be translated" and then TOTALLY forgot to do it. Thanks for picking these things up :)

Soean commented:
- The images should be deleted form the media library when deleting the block.
- The images should be attached to the post and deleted when deleting the post.

jasmussen commented:
Do we import the images into the media library when opengraph data is available, or is this only when scraping? If the former is the case, then perhaps it would be fine to start without the scraping and consider that as an additive enhancement to look at separately/in the future?

notnownikki commented:
@Soean 

> The images should be deleted form the media library when deleting the block
> The images should be attached to the post and deleted when deleting the post.

I didn't do these for one reason - caching. If the images are deleted when you delete the block, then if you go to preview the link again it will be referencing images that no longer exist. Also, I don't think there's a way to do things when a block is deleted at the moment...

@jasmussen Yes, we always import the images so we're not hotlinking images in posts, and only serve thumbnail size images. The scraping is already done and working!

notnownikki commented:
About the caching thing... the cache is already limited to the current blog, because it's referencing images in the blog's media library. Perhaps we get rid of caching altogether? I'd really like @iseulde's input on this.

notnownikki commented:
Rebased to pick up clientid and api fetch changes

notnownikki committed:
Link preview block

Provides OpenGraph and content scraping previews for any URL,
used as a fallback by the Embed block when a URL cannot be
handled by oEmbed.

notnownikki commented:
Rebased and updated for master.

notnownikki committed:
Lint fix

gziolo commented:
@karmatosed, @mapk and @melchoyce - what's the current status of this block? Should it be refreshed?

mapk commented:
I just tried testing this PR and got this error in the browser:

`Fatal error: Cannot declare class WP_REST_Blocks_Controller, because the name is already in use in /var/www/html/wp-content/plugins/gutenberg/lib/class-wp-rest-blocks-controller.php on line 0`

I also see some untracked files that get added when I switch to this branch.

```
packages/block-serialization-spec-parser/parser.js
packages/block-serialization-spec-parser/parser.php
```

If someone can walk me through this, I can run another test.

gziolo commented:
> I just tried testing this PR and got this error in the browser

I also see some merge conflicts issues. This PR needs to be updated with the latest changes in `master` branch. I'm adding label `State` to make it clear. @notnownikki do you plan to update this code? It looks like there is still some interestest to bring it to core.

notnownikki commented:
Not sure on the status of this one, I'm not working on WordPress now, so I'll leave this open in case anyone else wants to continue, or happy to close it if other efforts are ongoing for the functionality.

gziolo commented:
Let's close as non-member won't be able to update this PR. They can always recreate based on the work done here.
